core/pkg/eval/json_evaluator.go

```diff
@@ -24,7 +24,11 @@ import (
 	"go.uber.org/zap"
 )
 
-const SelectorMetadataKey = "scope"
+const (
+	SelectorMetadataKey = "scope"
+
+	contextFlagKey = "$flagd.__flag_key__"
```

Comment:
Two thoughts about this:

- I like your `$flagd.` proposal, I think it provides enough namespacing that we don't need the `__`... so `$flagd.flag_key` seems adequate.
- We could make `$flagd` a property, with nested values (just `flag_key` for now), so to access it in a rule, we'd do `{"var" : "$flagd.flag_key"}`. If we don't do it this way, I think we'd need to escape the `.` in the name, which would feel clumsy, I think.
  - If for some reason we don't want to use nesting, we should NOT use `.` as a delimeter. Maybe `__` instead.

Reply:
I agree with everything you said. Having these values accessible in a user's configuration would be very helpful. It may be interesting once we allow users to access nested properties to be able to do something like this:

```json
{
	"fractionalEvaluation": [
		{"cat": [{"var":"$flagd.flag_key"}, {"var":"user.email"}]},
		[ "red", 25 ],
		[ "blue", 25 ],
		[ "green", 25 ],
		[ "yellow", 25]
	]
}
```

Automatically setting the current timestamp and other properties sounds like a good idea. We should create a follow up issue to track the ideas and implementation.

Reply:
> I agree with everything you said. Having these values accessible in a user's configuration would be very helpful. It may be interesting once we allow users to access nested properties to be able to do something like this:
> 
> ```json
> {
> 	"fractionalEvaluation": [
> 		{"cat": [{"var":"$flagd.flag_key"}, {"var":"user.email"}]},
> 		[ "red", 25 ],
> 		[ "blue", 25 ],
> 		[ "green", 25 ],
> 		[ "yellow", 25]
> 	]
> }
> ```
> 
> Automatically setting the current timestamp and other properties sounds like a good idea. We should create a follow up issue to track the ideas and implementation.

I could be confused, but I think we'd do the `cat` operation ourselves, right? So the users would still just mention the prop they want to bucket on, and we'd `cat` it in the evaluator implementation. The only thing missing is the prop user's specify can't currently be a nested prop: `some.nested.prop".

Reply:
@craigpastro:

@beeme1mr and I were talking more about this, and we came up with a proposal we'd like your input on:

We're thinking about changing the `fractionalEvaluation` in a couple ways:

- Support a version that doesn't require an explicit bucketing prop, and by default uses `flag_key` + `targetingKey` as discussed. This would look like :

```json
{
	"fractionalEvaluation": [
		[ "red", 25 ],
		[ "blue", 25 ],
		[ "green", 25 ],
		[ "yellow", 25 ]
	]
}
```

Which would probably be what people want in 90% of use-cases
- Allow an a json logic expression for the first param to derive a custom bucketing value, if you dont want to do above, which would look like:

```json
{
	"fractionalEvaluation": [
		{"cat": [{"var":"$flagd.flag_key"}, {"var":"user.email"}]},
		[ "red", 25 ],
		[ "blue", 25 ],
		[ "green", 25 ],
		[ "yellow", 25]
	]
}
```

- change the name to `fractional` (`fractionalEvaluation` is redundant), and keep the current implementation as is, but deprecate it so we can remove it with 1.0.

Reply:
1. The first proposal sounds nice and simplifies things a bit, which is always nice. Where does `targetingKey` come from in this case? Will you need to specify it in `targeting`?
```
"targeting": {
  "key": "email",
  "fractionalEvaluation": [...]
}
```

2. This is also nice! In this case it means that `jsonlogic` will handle the logic as long as the correct stuff is added to the context, is that right?

Reply:
> I like your $flagd. proposal,
 
 That is @beeme1mr's proposal from https://github.com/open-feature/flagd/issues/843#issuecomment-1682718897.

Reply:
`targetingKey` is a property we can assume will be present in the context in this case (and if it's not, we can log a warning and default the rule). This is a property we can assume to be defined. It's something we mention in the OpenFeature spec: https://openfeature.dev/specification/sections/evaluation-context#requirement-311

Reply:
Ah, I didn't know that. Thank you. If we can assume that `targetingKey` is in the context, then we don't need to add it from the JSON evaluator do we? That is, this PR could be simplified quite a bit?

Reply:
I'm not sure you'll want to add this to this PR (though you can if you'd like).